pkg/sql/colflow: BenchmarkColBatchScan failed

pkg/sql/colflow.BenchmarkColBatchScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20958673?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20958673?buildTab=artifacts#./artifacts/microbench/experiment) on master @ [94a1e22436ad935d8c7f9524cb8fc0fd7b24faaa](https://github.com/cockroachdb/cockroach/commits/94a1e22436ad935d8c7f9524cb8fc0fd7b24faaa):

Fatal error:

```
panic: non-nil metadata from *colexecutils.CancelChecker: &{[] <nil> [] <nil> <nil> <nil> <nil> bytes_read:0 rows_read:20479 rows_written:0 index_rows_written:0 index_bytes_written:0 stage_id:0 kv_cpu_time:0  <nil> <nil>} [recovered, repanicked]

```
Stack: 

```
goroutine 3439 gp=0xc0059e34a0 m=16 mp=0xc000181808 [running]:
panic({0x559daa0?, 0xc0070c8160?})
	GOROOT/src/runtime/panic.go:802 +0x168 fp=0xc00232b218 sp=0xc00232b168 pc=0x4baba8
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*Context).Stop(0xc008d82808, {0x6a975e0, 0xabde000})
	pkg/sql/sem/eval/context.go:650 +0x8c fp=0xc00232b258 sp=0xc00232b218 pc=0x1f6e44c
github.com/cockroachdb/cockroach/pkg/sql/colflow_test.BenchmarkColBatchScan.func1.deferwrap1()
	pkg/sql/colflow/colbatch_scan_test.go:161 +0x28 fp=0xc00232b280 sp=0xc00232b258 pc=0x4c66a08
panic({0x559daa0?, 0xc0070c8160?})
	GOROOT/src/runtime/panic.go:783 +0x132 fp=0xc00232b330 sp=0xc00232b280 pc=0x4bab72
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.InternalError(...)
	pkg/sql/colexecerror/error.go:301
github.com/cockroachdb/cockroach/pkg/sql/colexecop.NextNoMeta({0x6a99930, 0xc00718a100})
	pkg/sql/colexecop/operator.go:546 +0xd0 fp=0xc00232b3a0 sp=0xc00232b330 pc=0x1e87db0
github.com/cockroachdb/cockroach/pkg/sql/colflow_test.BenchmarkColBatchScan.func1(0xc001436588)
	pkg/sql/colflow/colbatch_scan_test.go:197 +0x9f8 fp=0xc00232bf10 sp=0xc00232b3a0 pc=0x4c66938
testing.(*B).runN(0xc001436588, 0x1)
	GOROOT/src/testing/benchmark.go:219 +0x190 fp=0xc00232bfa0 sp=0xc00232bf10 pc=0x5be1b0
testing.(*B).run1.func1()
	GOROOT/src/testing/benchmark.go:245 +0x48 fp=0xc00232bfe0 sp=0xc00232bfa0 pc=0x5be828
runtime.goexit({})
	src/runtime/asm_amd64.s:1693 +0x1 fp=0xc00232bfe8 sp=0xc00232bfe0 pc=0x4c4141
created by testing.(*B).run1 in goroutine 27
	GOROOT/src/testing/benchmark.go:238 +0x90
```
<details><summary>Log preceding fatal error</summary>
<p>

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.60GHz
BenchmarkColBatchScan
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkColBatchScan4164337358
    test_log_scope.go:82: use -show-logs to present logs inline
BenchmarkColBatchScan/rows=16
BenchmarkColBatchScan/rows=16-32         	   22083	     51249 ns/op	   5.00 MB/s	    7825 B/op	      86 allocs/op
BenchmarkColBatchScan/rows=256
BenchmarkColBatchScan/rows=256-32        	    8960	    129438 ns/op	  31.64 MB/s	   27992 B/op	     128 allocs/op
BenchmarkColBatchScan/rows=4096
BenchmarkColBatchScan/rows=4096-32       	     993	   1175677 ns/op	  55.74 MB/s	  199259 B/op	     200 allocs/op
BenchmarkColBatchScan/rows=65536

I260111 17:04:06.175682 1 (gostd) rand.go:288  [-] 1  random seed: -5118599864099136554
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkColBatchScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58598

--- a/pkg/sql/colexec/ordered_synchronizer.eg.go
+++ b/pkg/sql/colexec/ordered_synchronizer.eg.go
@@ -69,6 +69,9 @@ type OrderedSynchronizer struct {
 	// continueBatch, if set, indicates that the current output batch should be
 	// continued.
 	continueBatch bool
+	// tempBufferedMeta contains temporarily buffered metadata until the next
+	// input batch from the minimum input is fetched.
+	tempBufferedMeta []execinfrapb.ProducerMetadata
 }
 
 var (
@@ -115,8 +118,23 @@ func NewOrderedSynchronizer(
 	return os
 }
 
+// maybeEmitMeta checks whether the synchronizer has temporarily buffered any
+// metadata and returns one object if so.
+func (o *OrderedSynchronizer) maybeEmitMeta() *execinfrapb.ProducerMetadata {
+	if len(o.tempBufferedMeta) == 0 {
+		return nil
+	}
+	meta := o.tempBufferedMeta[0]
+	o.tempBufferedMeta[0] = execinfrapb.ProducerMetadata{}
+	o.tempBufferedMeta = o.tempBufferedMeta[1:]
+	return &meta
+}
+
 // Next is part of the Operator interface.
 func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetadata) {
+	if meta := o.maybeEmitMeta(); meta != nil {
+		return nil, meta
+	}
 	if !o.heapInitialized {
 		for i := range o.inputs {
 			if o.inputBatches[i] != nil {
@@ -148,6 +166,10 @@ func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetada
 	}
 	o.continueBatch = false
 	for batchDone := false; !batchDone; {
+		if meta := o.maybeEmitMeta(); meta != nil {
+			o.continueBatch = true
+			return nil, meta
+		}
 		if o.advanceMinBatch {
 			// Advance the minimum input batch, fetching a new batch if
 			// necessary.
@@ -155,12 +177,19 @@ func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetada
 			if o.inputIndices[minBatch]+1 < o.inputBatches[minBatch].Length() {
 				o.inputIndices[minBatch]++
 			} else {
-				batch, meta := o.inputs[minBatch].Root.Next()
-				if meta != nil {
-					o.continueBatch = true
-					return nil, meta
+				// We're in the middle of updating the heap, so we need to fetch
+				// the next batch and update the comparators before we can emit
+				// any metadata. Thus, we'll temporarily buffer all meta and
+				// emit it at a convenient point in time.
+				for {
+					batch, meta := o.inputs[minBatch].Root.Next()
+					if meta != nil {
+						o.tempBufferedMeta = append(o.tempBufferedMeta, *meta)
+						continue
+					}
+					o.inputBatches[minBatch] = batch
+					break
 				}
-				o.inputBatches[minBatch] = batch
 				o.inputIndices[minBatch] = 0
 				o.updateComparators(minBatch)
 			}
@@ -319,7 +348,11 @@ func (o *OrderedSynchronizer) Init(ctx context.Context) {
 }
 
 func (o *OrderedSynchronizer) DrainMeta() []execinfrapb.ProducerMetadata {
-	var bufferedMeta []execinfrapb.ProducerMetadata
+	// We generally don't expect any temporarily buffered metadata here, but in
+	// some edge cases it's possible - guarantee that it's emitted at least
+	// during DrainMeta.
+	bufferedMeta := o.tempBufferedMeta
+	o.tempBufferedMeta = nil
 	if o.span != nil {
 		for i := range o.inputs {
 			for _, stats := range o.inputs[i].StatsCollectors {

--- a/pkg/sql/colexec/ordered_synchronizer_tmpl.go
+++ b/pkg/sql/colexec/ordered_synchronizer_tmpl.go
@@ -89,6 +89,9 @@ type OrderedSynchronizer struct {
 	// continueBatch, if set, indicates that the current output batch should be
 	// continued.
 	continueBatch bool
+	// tempBufferedMeta contains temporarily buffered metadata until the next
+	// input batch from the minimum input is fetched.
+	tempBufferedMeta []execinfrapb.ProducerMetadata
 }
 
 var (
@@ -135,8 +138,23 @@ func NewOrderedSynchronizer(
 	return os
 }
 
+// maybeEmitMeta checks whether the synchronizer has temporarily buffered any
+// metadata and returns one object if so.
+func (o *OrderedSynchronizer) maybeEmitMeta() *execinfrapb.ProducerMetadata {
+	if len(o.tempBufferedMeta) == 0 {
+		return nil
+	}
+	meta := o.tempBufferedMeta[0]
+	o.tempBufferedMeta[0] = execinfrapb.ProducerMetadata{}
+	o.tempBufferedMeta = o.tempBufferedMeta[1:]
+	return &meta
+}
+
 // Next is part of the Operator interface.
 func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetadata) {
+	if meta := o.maybeEmitMeta(); meta != nil {
+		return nil, meta
+	}
 	if !o.heapInitialized {
 		for i := range o.inputs {
 			if o.inputBatches[i] != nil {
@@ -168,6 +186,10 @@ func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetada
 	}
 	o.continueBatch = false
 	for batchDone := false; !batchDone; {
+		if meta := o.maybeEmitMeta(); meta != nil {
+			o.continueBatch = true
+			return nil, meta
+		}
 		if o.advanceMinBatch {
 			// Advance the minimum input batch, fetching a new batch if
 			// necessary.
@@ -175,12 +197,19 @@ func (o *OrderedSynchronizer) Next() (coldata.Batch, *execinfrapb.ProducerMetada
 			if o.inputIndices[minBatch]+1 < o.inputBatches[minBatch].Length() {
 				o.inputIndices[minBatch]++
 			} else {
-				batch, meta := o.inputs[minBatch].Root.Next()
-				if meta != nil {
-					o.continueBatch = true
-					return nil, meta
+				// We're in the middle of updating the heap, so we need to fetch
+				// the next batch and update the comparators before we can emit
+				// any metadata. Thus, we'll temporarily buffer all meta and
+				// emit it at a convenient point in time.
+				for {
+					batch, meta := o.inputs[minBatch].Root.Next()
+					if meta != nil {
+						o.tempBufferedMeta = append(o.tempBufferedMeta, *meta)
+						continue
+					}
+					o.inputBatches[minBatch] = batch
+					break
 				}
-				o.inputBatches[minBatch] = batch
 				o.inputIndices[minBatch] = 0
 				o.updateComparators(minBatch)
 			}
@@ -266,7 +295,11 @@ func (o *OrderedSynchronizer) Init(ctx context.Context) {
 }
 
 func (o *OrderedSynchronizer) DrainMeta() []execinfrapb.ProducerMetadata {
-	var bufferedMeta []execinfrapb.ProducerMetadata
+	// We generally don't expect any temporarily buffered metadata here, but in
+	// some edge cases it's possible - guarantee that it's emitted at least
+	// during DrainMeta.
+	bufferedMeta := o.tempBufferedMeta
+	o.tempBufferedMeta = nil
 	if o.span != nil {
 		for i := range o.inputs {
 			for _, stats := range o.inputs[i].StatsCollectors {

--- a/pkg/sql/colflow/colbatch_scan_test.go
+++ b/pkg/sql/colflow/colbatch_scan_test.go
@@ -22,7 +22,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/fetchpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecargs"
-	"github.com/cockroachdb/cockroach/pkg/sql/colexecop"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
@@ -32,6 +31,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
 )
 
 // TestColBatchScanMeta makes sure that the ColBatchScan propagates the leaf
@@ -194,7 +194,13 @@ func BenchmarkColBatchScan(b *testing.B) {
 				b.StartTimer()
 				tr.Init(ctx)
 				for {
-					bat := colexecop.NextNoMeta(tr)
+					bat, meta := tr.Next()
+					if meta != nil {
+						if meta.Metrics != nil {
+							continue
+						}
+						b.Fatal(errors.AssertionFailedf("unexpected meta: %+v", meta))
+					}
 					if bat.Length() == 0 {
 						break
 					}

